File: Projects/MetadataRetrieverPlatform/app/src/main/java/com/example/metadataretrieverplatform/RetrieverTestToni.java
package com.example.metadataretrieverplatform;

import static android.media.MediaMetadataRetriever.METADATA_KEY_HAS_VIDEO;

import android.app.Activity;
import android.content.Context;
import android.media.MediaMetadataRetriever;
import android.util.Log;

import androidx.annotation.OptIn;
import androidx.media3.common.MediaItem;
import androidx.media3.common.util.UnstableApi;
import androidx.media3.common.util.Util;
import androidx.media3.datasource.DefaultDataSource;
import androidx.media3.exoplayer.MetadataRetriever;
import androidx.media3.exoplayer.source.ProgressiveMediaSource;

import java.io.IOException;
import java.util.concurrent.ExecutionException;

public class RetrieverTestToni {
    public static void startMetadataRetrievalTest(Activity context) {
        String testPath = "/sdcard/Download/mediadataset/dizzy.mp4";
        if (Util.maybeRequestReadStoragePermission(context, MediaItem.fromUri(testPath))) {
            return;
        }
        // Run each version once to avoid class loading etc to influence runtime.
        retrieveMetadataPlatform(testPath);
//        retrieveMetadataMedia3(context, testPath);
        // Run 50 times each and get the average.
        long platformTotal = 0;
        long media3Total = 0;
        for (int i = 0; i < 100; i++) {
            platformTotal += retrieveMetadataPlatform(testPath);
//            media3Total += retrieveMetadataMedia3(context, testPath);
        }
        Log.d("tonihei", "(single) platform: " + (platformTotal / 50) + ", media3: " + (media3Total / 50));
    }


    private static long retrieveMetadataPlatform(String path) {
        long startTimeNs = System.nanoTime();
        try (MediaMetadataRetriever mediaMetadataRetriever = new MediaMetadataRetriever()) {
            mediaMetadataRetriever.setDataSource(path);
            mediaMetadataRetriever.extractMetadata(METADATA_KEY_HAS_VIDEO);
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return (System.nanoTime() - startTimeNs) / 1000;
    }

//    @OptIn(markerClass = UnstableApi.class)
//    private static long retrieveMetadataMedia3(Context context, String path) {
//        long startTimeNs = System.nanoTime();
//        try {
//            MetadataRetriever.retrieveMetadata(
//                            new ProgressiveMediaSource.Factory(new DefaultDataSource.Factory(context)),
//                            MediaItem.fromUri(path))
//                    .get();
//        } catch (ExecutionException | RuntimeException | InterruptedException e) {
//            throw new RuntimeException(e);
//        }
//        return (System.nanoTime() - startTimeNs) / 1000;
//    }

}

File: Projects/Retriever/RetrieverMedia3Toni/app/src/main/java/com/example/retrievermedia3toni/RetrieverTestToni.kt
package com.example.retrievermedia3toni

import android.app.Activity
import android.content.Context
import android.os.Trace
import androidx.annotation.OptIn
import androidx.media3.common.MediaItem
import androidx.media3.common.util.Log
import androidx.media3.common.util.UnstableApi
import androidx.media3.common.util.Util
import androidx.media3.datasource.DefaultDataSource
import androidx.media3.exoplayer.MetadataRetriever
import androidx.media3.exoplayer.source.ProgressiveMediaSource
import androidx.media3.exoplayer.source.TrackGroupArray
import java.util.Locale
import java.util.concurrent.Future

@OptIn(UnstableApi::class)
class RetrieverTestToni(
    private val mediaPath: String, private val numIter: Int
) {

    fun startMetadataRetrievalTestSerial(context: Context): Long {
        var totalTimeUs: Long = 0L
        for (i in 0..<numIter) {
            val currTimeUs = retrieveMetadataMedia3(context, i)
            totalTimeUs += currTimeUs
//            Log.d(
//                TAG, String.format(
//                    Locale.ROOT, "%30s %3d: %7d", mediaPath, i, currTimeUs
//                )
//            )
        }
        val meanTimeUs = totalTimeUs / numIter
        printMeanRetrieverTime(meanTimeUs, "Serial")
        return meanTimeUs
    }

    fun startMetadataRetrievalTestBulk(context: Context): Long {
        val totalTimeUs = retrieveMetadataMedia3Bulk(context)
        val meanTimeUs = totalTimeUs / numIter
        printMeanRetrieverTime(meanTimeUs, "Bulk")
        return meanTimeUs
    }

    private fun printMeanRetrieverTime(meanTimeUs: Long, mode: String) {
        Log.d(
            TAG, String.format(
                Locale.ROOT,
                "[$TAG][$mode] Mean Retriever Time: %30s %7d",
                mediaPath,
                meanTimeUs,
            )
        )
    }

    private fun retrieveMetadataMedia3(context: Context, i: Int): Long {
        Trace.beginSection("retrieveMetadataMedia3Aman $i")
        val startTimeNs = System.nanoTime()

        try {
            val mediaSourceFactory =
                ProgressiveMediaSource.Factory(DefaultDataSource.Factory(context))
            val mediaItem = MediaItem.fromUri(mediaPath)
            MetadataRetriever.retrieveMetadata(mediaSourceFactory, mediaItem).get()
        } catch (e: Exception) {
            throw RuntimeException(e)
        }

        Trace.endSection()
        return (System.nanoTime() - startTimeNs) / 1000
    }

    private fun retrieveMetadataMedia3Bulk(context: Context): Long {
        val startTimeNs = System.nanoTime()

        val futures = mutableListOf<Future<TrackGroupArray>>()
        for (i in 0..<numIter) {
            try {
                val mediaSourceFactory =
                    ProgressiveMediaSource.Factory(DefaultDataSource.Factory(context))
                val mediaItem = MediaItem.fromUri(mediaPath)
                val future = MetadataRetriever.retrieveMetadata(mediaSourceFactory, mediaItem)
                futures.add(future)
//              Log.d(TAG, "$i: Retriever submitted.")
            } catch (e: Exception) {
                Log.e(TAG, "$i: Error retrieving metadata: $e")
            }
        }
        val numRetrievers = futures.size
        for (i in 0..<numRetrievers) {
            try {
                futures[i].get()
//                Log.d(TAG, "$i: Future completed for retriever.")
            } catch (e: Exception) {
                Log.e(TAG, "$i: Error waiting for future of retriever to complete", e)
            }
        }

        return (System.nanoTime() - startTimeNs) / 1000
    }

}